enhancements/baremetal-config.md

```diff
@@ -0,0 +1,153 @@
+---
+title: Baremetal IPI Config Enhancement
+authors:
+  - "@sadasu"
+reviewers:
+  - "@deads2k"
+  - "@hardys"
+  - "@enxebre"
+  - "@squeed"
+  - "@abhinavdahiya"
+
+approvers:
+  - "@deads2k"
+
+creation-date: 2019-10-23
+last-updated: 2019-10-23
+status: implemented
+see-also:
+  - "https://github.com/openshift/api/pull/480"  
+replaces:
+  - 
+superseded-by:
+  - 
+---
+
+# Config required for Baremetal IPI deployments
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions [optional]
+
+None.
+
+## Summary
+
+The configuration required to bring up a Baremetal IPI (metal3) cluster needs
+to be availabe in a Config Resource. The installer aleady adds some configuration
+to the Infrastructure Config CR. It is also aware of the additional config
+required during the deployment of a metal3 cluster. This enhancement proposes
+to extend the BareMetaPlatformStatus CR to hold these additional values needed
+for a metal3 deployment.
+
+## Motivation
+
+Adding these additional config items to the BareMetalPlatformStatus CR, allows
+versioning of these config items.
+
+### Goals
+
+The goal of this enhancement is to enable the installer to add vital config
+values for a Baremetal IPI deployment to a CR. The MAO will then be able to
+access these values when it kicks off a metal3 cluster.
+
+Update lifecycle for bootimage is an unsolved problem for all platforms. For
+more information, refer to : https://github.com/openshift/os/issues/381.
+
+### Non-Goals
+
+All config values except the RHCOSImageUrl are not expected to change after a
+metal3 cluster is up. Handling changes to this value after install time has
+not been explored and that is a non-goal at this time.
+
+### Proposal
+
+To support configuration of a Baremetal IPI (metal3) deployment, a set of new
+config items are required to be made availabe via a Config Resource. The
+BareMetalPlatformStatus is an ideal fit because it contains configuration
+that is specific to a Baremetal deployment.
+
+This proposal aims to add the following to BareMetalPlatformStatus:
+
```

Comment:
ProvisioningInterface, ProvisioningDHCPRange,  ProvisioningIP, ProvisioningNetworkCIDR seem like the information that best belongs on the baremetal host CR itself, because my assumption was that it described a baremetal-server

Reply:
They aren't related to the BareMetalHost CR - that describes the hosts to be provisioned, this is an implementation detail of how the provisioning service works

Reply:
Right. The host CR is the inventory that the installer uses to fulfill a Machine. These settings tell the provisioning system how to do that work, by telling it which network it can use and what IPs it should use on that network.

Reply:
This could go either way.  For example, you'd want it on the Machine if you wanted the machine-API to be able to support multiple providers (e.g. "make machine A using provider A and make machine B using provider B").  If we don't want to support that (and we don't for other providers like AWS), then having these configured once in a central location like this makes sense.  And even if we do support per-machine settings, having a central default makes sense to support folks creating Machines who don't want to care about selecting these things.  We have precedent for all of these, like [the default AWS region on the Infrastructure object][1], which can be overridden in more-specific configs as desired, [the GCP project ID on the Infrastructure object][2], which maybe can't be overridden (I dunno), and on libvirt the libvirtd URI is [on the machine provider][3] with [no cluster-level default][4].  Also, if this is going to be a machine-API-provider-only config, not used by other cluster components, it might be worth putting the cluster default somewhere specific to the machine-API provider, and not in the more generic Infrastructure.

[1]: https://github.com/openshift/api/blob/d0b31d707c464221d1eb24846b0d0bbe57040102/config/v1/types_infrastructure.go#L140-L141
[2]: https://github.com/openshift/api/blob/d0b31d707c464221d1eb24846b0d0bbe57040102/config/v1/types_infrastructure.go#L157-L158
[3]: https://github.com/openshift/cluster-api-provider-libvirt/blob/c32e2ea813cf282e2d85e24e882ed3e2e897f500/pkg/apis/libvirtproviderconfig/v1beta1/libvirtmachineproviderconfig_types.go#L35
[4]: https://github.com/openshift/api/blob/d0b31d707c464221d1eb24846b0d0bbe57040102/config/v1/types_infrastructure.go#L113-L135

Reply:
@wking Thanks for the comments. 
Machine-api-operator does refer to the PlatformStatus CR which contains the BareMetalPltaformStatus.

@enxebre is there a different CR that the machine-api-operator refers to and is more appropriate for this?

Reply:
> And even if we do support per-machine settings, having a central default makes sense to support folks creating Machines who don't want to care about selecting these things. 

I think that is the key issue here, the provisioning toolchain is supposed to be an implementation detail, so we do need some way to set sane defaults, and allow them to be overridden easily by the installer to match local network configuration.

It *could* be one day that we need custom configuration per machineset or something (I'm thinking a spine/leaf deployment topology), but it's not something currently under consideration, we just need to wire in the defaults from the installer as a first step.